R/keypair.R

```diff
@@ -0,0 +1,92 @@
+# Support for Snowflake key-pair authentication for SPCS and the REST API.
+keypair_credentials <- function(
+  account,
+  user,
+  private_key,
+  spcs_endpoint = NULL,
+  role = "PUBLIC"
+) {
+  jwt <- generate_jwt(account, user, private_key)
+  # Important: the SPCS ingress handles key-pair authentication *differently*
+  # than the REST API. In particular, we can't pass the signed JWT as a bearer
+  # token, we have to go through an exchange step.
+  if (is.null(spcs_endpoint)) {
+    return(list(
+      Authorization = paste("Bearer", jwt),
+      # Snowflake requires this additional header for JWTs, presumably to
+      # distinguish them from OAuth access tokens.
+      `X-Snowflake-Authorization-Token-Type` = "KEYPAIR_JWT"
+    ))
+  }
+  account_url <- sprintf("https://%s.snowflakecomputing.com", account)
+  token <- exchange_jwt_for_token(account_url, jwt, spcs_endpoint, role)
+  # Yes, this is actually the format of the Authorization header that SPCS
+  # requires.
+  return(list(
+    Authorization = sprintf('Snowflake Token="%s"', token)
+  ))
+}
+
+# Generate a JWT that can be used for Snowflake "key-pair" authentication.
+generate_jwt <- function(account, user, private_key, iat = NULL, jti = NULL) {
+  check_installed("jose", "for key-pair authentication")
```

Comment:
Ah, you've already set up `jose` as an optional dependency. So it should be in `Suggests` not `Imports`, and if that makes sense here, it probably makes sense for `RcppTOML` too.